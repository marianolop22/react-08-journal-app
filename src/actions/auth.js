import { types } from "../types/types";
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { notesLogoutCleaning } from './notes'

//esto es una accion asincrona
export const startLoginEmailPassword = (email, password) => {
    return ( dispatch ) => {

        dispatch ( startLoading() );

        //se agrega el return para que ser puedan hacer las pruebas
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ({user}) => {
                dispatch (login (user.uid, user.displayName) );
            })
            .catch( err => {
                Swal.fire(
                    'Fail',
                    err.message,
                    'error'
                );
            })
            .finally( ()=> {
                dispatch ( finishLoading()  );
            })
    }
};

export const startRegisterWithEmailPasswordName = (email,  password, name) => {
    
    return ( dispatch ) => {
        return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async ({ user }) => {
                await user.updateProfile({displayName: name});
                dispatch ( login ( user.uid, user.displayName ));
            })
            .catch( err => {
                Swal.fire(
                    'Fail',
                    err.message,
                    'error'
                );
            })
    }
};


export const startGoogleLogin = () => {

    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch ( login ( user.uid, user.displayName ));
            })
    }
};



export const startLogout = () => {
    return async ( dispatch ) => {

        // dispatch ( startLoading() );

        await firebase.auth().signOut()
        dispatch ( logout () );
        dispatch ( notesLogoutCleaning ());
        // dispatch ( finishLoading() );
    }
};

export const login = (uid, displayName ) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
    })

export const logout = () => ({
    type: types.logout
})