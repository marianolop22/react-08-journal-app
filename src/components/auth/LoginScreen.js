import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {

    //esto te permite ejecutar la funcion dispatch de lo que tengo en el store
    //redux, solito, busca que dispatch debe hacer
    const dispatch = useDispatch(); 
    const { ui: {loading}  } = useSelector( state => state );
    
    const [formValues, handleInputChange] = useForm ({
        email:'nando@pepe.com.ar',
        password:'123456'
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch ( startLoginEmailPassword ( email, password) );
    };


    const hangleGoogleLogin = () => {
        dispatch ( startGoogleLogin());
    };

    return (
        <>
            <h3 className="auth__title">Login</h3>

            {/* {
                    (msgError) &&
                    (
                        <div className="auth__alert-error">
                            { msgError }
                        </div>
                    )
            } */}

            <form onSubmit={handleLogin}>
                <input
                    className="auth__input"
                    type="text"
                    placeholder="Email"
                    name="email"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input
                    className="auth__input"
                    type="password"
                    placeholder="Password"
                    name="password"
                    value={password}
                    onChange={handleInputChange}
                />

                <button
                    className="btn btn-primary btn-block"
                    type="submit"
                    disabled={loading}
                >
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div 
                        className="google-btn"
                        onClick={hangleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link className="link" to="/auth/register">
                    Create new account
                </Link>

            </form>
      </>
    )
}