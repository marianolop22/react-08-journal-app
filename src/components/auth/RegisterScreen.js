import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { setError, removeError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from "../../actions/auth";


export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { ui: {msgError}  } = useSelector( state => state );

    const [formValues, handleInputChange] = useForm ({
        name: 'nando',
        email:'nando@pepe.com.ar',
        password:'123456',
        password2:'123456'
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch (  startRegisterWithEmailPasswordName(email, password, name) );
        } 
    };

    const isFormValid = () => {

        if (name.trim().length === 0 ) {
            dispatch ( setError ('Debe ingresar un nombre'));
            return false;
        } else if ( !validator.isEmail( email ) ) {
            dispatch ( setError ('Email invalido'));
            return false;
        } else if ( password !== password2 || password < 5) {
            dispatch ( setError ('Los pass deben ser iguales y de mas de  6  caracteres'));
            return false;
        }

        dispatch ( removeError ());
        return true;    
    };




    return (
        <>
            <h3 className="auth__title">Register</h3>  
            <form onSubmit={handleRegister}>

                {
                    (msgError) &&
                    (
                        <div className="auth__alert-error">
                            { msgError }
                        </div>
                    )
                }

                <input
                    className="auth__input"
                    type="text"
                    placeholder="Name"
                    name="name"
                    autoComplete="off"
                    value={name}
                    onChange={handleInputChange}
                />

                <input
                    className="auth__input"
                    type="text"
                    placeholder="Email"
                    name="email"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input
                    className="auth__input"
                    type="password"
                    placeholder="Password"
                    name="password"
                    value={password}
                    onChange={handleInputChange}
                />
                <input
                    className="auth__input"
                    type="password"
                    placeholder="Confirm Password"
                    name="password2"
                    value={password2}
                    onChange={handleInputChange}
                />

                <button
                    className="btn btn-primary btn-block mb-5"
                    type="submit"
                >
                    Register
                </button>

                <Link className="link" to="/auth/login">
                    Already registered?
                </Link>

            </form>
        </>
    )
}
