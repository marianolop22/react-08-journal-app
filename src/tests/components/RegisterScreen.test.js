import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';

import { MemoryRouter } from "react-router-dom";
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk'; 
import { RegisterScreen } from '../../components/auth/RegisterScreen';
import { types } from '../../types/types';

//todo esto es para generar un estado para poder utilizar el redux
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore (initState);

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen/>
        </MemoryRouter>
    </Provider>
);


describe ('Pruebas de RegisterScreen', ()=> {

    test('debe de renderizar', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de tirar error en input', () => {

        const emailField = wrapper.find('input[name="email"]');

        emailField.simulate('change', {
            target: {
                value: '',
                name:'email'
            }
        });

        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });

        const actions = store.getActions();
        // console.log(actions);

        expect( actions[0] ).toEqual({
            type:types.uiSetError,
            payload: 'Email invalido'
        });
    });

    test('debe de mostrar la caja de alerta con un error', () => {
        
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email no es correcto'
            }
        }
        
        let store = mockStore (initState);
        
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen/>
                </MemoryRouter>
            </Provider>
        );
        
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError);

    })
    


});