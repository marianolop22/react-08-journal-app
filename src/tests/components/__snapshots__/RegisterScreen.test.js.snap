// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de RegisterScreen debe de renderizar 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="nando"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="nando@pepe.com.ar"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
